### ISSUE TITLE
Implement Announcements for community moderators

### ISSUE THREAD

**Issue by ori-near** [NEAR-DevHub/neardevhub-bos#540]
# Problem
Right now, it's hard for users to stay informed and engaged across our various communities. The existing homepage/feed is all about proposals and contributor reports, which is great, but it doesn't give people the full picture of what's happening.

# Solution
Turn the existing homepage into an announcement feed, curated by community moderators. This way, we can share all the latest updates in one place.

# User Stories
- As a community admin, I want to be able to share the latest announcements with the community so that community members can get the latest updates.
- As a user who cares about specific communities, I want to see the latest announcements and engage with them. 

# Acceptance Criteria

**DevHub community page requirements:**
- [ ] There is a new Announcements tab, which is the default tab that users see when they land on a given community
- [ ] Community Admins see a share update input text area field (e.g. What's happening)
  - [ ] Non community admins do not see the What's happening share box, since they cannot post announcements
  - [ ] The share update input text area field contains all the same elements from the near.org, including supporting markdown, image update, as well as the preview.
  - [ ] Community admins can type their update, preview it, upload image, and click post. Upon post, the update shows up in the announcements section below using the community profile image and name of that community.
- [ ] All non-admin users see the announcements section. 
  - [ ] The feed is sorted by latest, but users can also filter by recent comments
  - [ ] Users can click Load More to load additional posts
  - [ ] Users can interact with announcements posts (like, add comment, copy link to post, or share).
  - [ ] Users can report posts (clicking the three dots). _Note: @ailisp will handle this part in a separate ticket. Please check in with him._
  - [ ] While the original parent post is posted under the community picture and name, if a community admin adds a comment or interacts with any posts, their response is posted using their individual community profile image and name.

**near.org and near social requirements:**
- [ ] Any announcements posted on DevHub communities will also show up on near.org and near social gateways

**Backend storage requirements:**
- [ ] Use Social DB and follow near.org conventions-

# Mockups
https://www.figma.com/file/6bdpB5IrHztN7E40Rjle1N/Community-Feature%3A-Announcements?type=design&node-id=12%3A6010&mode=design&t=M1jwBsC4NvLrn35p-1

# Technical Details
@frol to add some considerations, including interactions with Social DB + feed/notification details. Feel free to update any of the acceptance criteria above as well.

**Comment by frol** [comment]
# Technical Details

Technical requirements:

* Announcement and discussion posts on DevHub should be visible on the regular feed of NEAR Social
* Ideally, we should plug into existing NEAR Social conventions, so existing gateways won't need to be changed

The solution:

Each community should have 2 NEAR subaccounts (naming is too verbose, so suggestions are welcome):

1. `<community-handle>.communities.devhub.near` - for announcement posts
2. `discussions.<community-handle>.communities.devhub.near` - for discussion posts

With such a setup, NEAR Social users will be able to subscribe to those regular accounts like to any regular user and have granular control over all the posts in their feeds and also like, reshare, and comment "natively".

These accounts don't even need access keys, upon setup, we need to grant SocialDB access to `devhub.near` account, so `devhub.near` contract can call `set("<community-handle>.devhub.near/post/main", ...)` on `social.near`. However, we would need a small smart contract to set things up, and we should add some recovery full access key, just in case.

**User-flow when posting to discussions:** To post to discussions, we need to automate users creating a post on their own NEAR Social feed and resharing to the `discussions.<community-handle>.communities.devhub.near`. As such, the post will be visible on NEAR Social feed, authored by the original author, and visible on the `discussions.<community-handle>.communities.devhub.near` account feed as a reshared post.

**User-flow when posting to announcements:** First of all, only community moderators should be allowed to post there. To post to announcements, we need to create a post on behalf of the `<community-handle>.communities.devhub.near` NEAR Social feed and offer to reshare it to the personal account. As such, the announcement posts will be visible on NEAR Social `<community-handle>.communities.devhub.near` account feed.

Other considerations:

1. **Moderation:** We may implement pre-moderation with the posts staging in the `devhub.near` contract first before being reshared by `discussions.*` account. We may also consider post-moderation by filtering out uninvited content.
2. **Editing:** I think editing is not critical for discussions and announcements and using native NEAR Social Feed we may expect that NEAR Social will eventually come up with some conversion on editing posts and DevHub will just inherit it.
3. **Nested discussions:** One level of nesting is mostly enough for now. Freeform nesting is as hard to follow as no nesting in some cases, so we need to strike some balance.

**Comment by frol** [comment]
# UI

I suggest reusing existing ActivityFeed components from near.org as much as possible. Since we will plug into SocialDB natively, the view feed component can be taken from the Profile page (https://near.org/near/widget/ProfilePage?accountId=devgovgigs.near).

**Comment by elliotBraem** [comment]
@frol I think we could simplify this a bit.

As I understand, the purpose of `discussions.<community-handle>.communities.devhub.near` is just for being able to follow, and so the discussion posts appear in the user's feed. Moderation looks to be handled by @ailisp's [Moderation 2: SocialDB schema and Indexers to support moderation](https://github.com/near/neardevhub-widgets/issues/541). The discussions aren't going to actually go through the discussions account; it's just reposting for exposure -- likes, comments, etc _should_ occur on the original creator's post, but I think you'll find this difficult to implement in existing social feeds.

I don't think we need to do any reposting-- in fact, I think cross posting would create a lot of clutter and inconsistencies. Instead, I'd like to do a spike demonstrating personalized feeds and discussion post indexing. 


Regarding announcements, I'm wondering how we plan to sync permissions between the devhub contract and the announcements account? I also think there is a bottle neck in that the community moderators have to be the ones to create the announcement... why not deploy a sputnik contract and allow function call proposals to be voted on by the DAO? Then moderators can create posts, but community members can propose posts, too? [Here's an example](https://near.social/hack.near/widget/DAO.Social) of this already implemented, and there are other ecosystem components that can be used, too.

**Comment by frol** [comment]
> it's just reposting for exposure
> ...
> I don't think we need to do any reposting-- in fact, I think cross posting would create a lot of clutter and inconsistencies. 

@elliotBraem You missed the fact that I suggest using resharing, not reposting. Resharing is only supported on near.social, but I confirmed that near.org will also add the support. All likes and comments get to the original post.

<img width="907" alt="image" src="https://github.com/near/neardevhub-widgets/assets/304265/f3701f58-3388-484c-bd6a-4cef974cf9d2">

https://near.social/mob.near/widget/MyPage?accountId=dcjanio.near

> Regarding announcements, I'm wondering how we plan to sync permissions between the devhub contract and the announcements account? 

This is where a small contract deployed to the `<community-handle>.communities.devhub.near` will be handy. It will need to call [`grant_write_permission({predecessor_id: "devgovgigs.near"})`](https://github.com/NearSocial/social-db/blob/d28c647252ce25a06c70c3b7f4930ccdcd217fd9/contract/src/permission.rs#L29-L34) on Social DB, so `devgovgigs.near` contract will be able to call `set("<community-handle>.communities.devhub.near/post/main")` on SocialDB.

> why not deploy a sputnik contract and allow function call proposals to be voted on by the DAO? 

This is an advanced mode we may introduce in the future. At the current stage, it would only introduce an overhead and slow down the development.

**Comment by frol** [comment]
> All right, I didn't get that in the beginning. You want to have a small smart contract deployed on each community subaccount in order to grant permissions on socialdb.

@petersalomonsen Yes, and potentially some other setup tasks.

> I assume this is a contract that can not and will not be upgraded at any time? And if so, where to store the keys in a safe way? Something that community admins needs to keep?

The contract may implement self_upgrade method just like devhub’s contract has. The full access public key will be hardcoded in devhub contract. Ori and I can store the private key for the time being.

> Unless that smart contract contains a JS scripting engine, and updates can be done by submitting scripts to it through the devhub contract.

Rust contracts can also be redeployed from the contract itself, but it is riskier as we may redeploy an invalid Wasm and cut ourselves out of contract methods, hence I would still want to keep a full access key for disaster recovery.

**Comment by ailisp** [comment]
Adding myself for the contract part.

**Comment by elliotBraem** [comment]
Adding @Megha-Dev-19 for the front end stylings.

We just discussed and she will be bootstrapping a basic feed based on the Figma provided.

**Comment by ailisp** [comment]
> @elliotBraem You missed the fact that I suggest using resharing, not reposting. Resharing is only supported on near.social, but I confirmed that near.org will also add the support. All likes and comments get to the original post.

@frol I'm confused here, I only see "Repost" and "Share" on near social, and share are sharing by copy link, send link to email or twitter. What is the "resharing" you are refer to?

<img width="196" alt="image" src="https://github.com/near/neardevhub-widgets/assets/13259400/5310ad24-def1-4a1d-97c3-12156f3f9c45">

<img width="185" alt="image" src="https://github.com/near/neardevhub-widgets/assets/13259400/c57426d2-6c15-4d24-8d08-b3a99261bc07">

**Comment by ailisp** [comment]
> This is where a small contract deployed to the <community-handle>.communities.devhub.near will be handy. It will need to call [grant_write_permission({predecessor_id: "devgovgigs.near"})](https://github.com/NearSocial/social-db/blob/d28c647252ce25a06c70c3b7f4930ccdcd217fd9/contract/src/permission.rs#L29-L34) on Social DB, so devgovgigs.near contract will be able to call set("<community-handle>.communities.devhub.near/post/main") on SocialDB.

I understand this part and the discussion about contract self upgrade. But I don't think we need resharing to implement the ticket's requirement. Considering these requirements:

>  While the original parent post is posted under the **community picture and name**, if a community admin adds a comment or interacts with any posts, their response is posted using their **individual community profile image and name**.

>  Any announcements posted on DevHub communities will also show up on near.org and near social gateways

The top level announcement posts are set `set("<community-handle>.communities.devhub.near/post/main")` on SocialDB from a community admin account. It's a social db post so we will automatically get "also show up on near.org and near social". Then later comments from community admins, are posted with their individual profile image and name. (The statement here is a bit confusing since it says "individual community profile image", but I talked with Ori and I can confirm here means individual profile image). So, the comment UX will be exactly same as comment a normal social db post and no additional work required here.

**Comment by frol** [comment]
> I'm confused here, I only see "Repost" and "Share" on near social, and share are sharing by copy link, send link to email or twitter. What is the "resharing" you are refer to?

Indeed, the button is called "repost", but it is not cloning the post as our current devhub contract does when we "cross post" DevHub posts to Social Feed. There is no such button or near.org, though and the only way there is to actually copy-paste and post again ("repost"). I want us to re-use near.social style of reposting.

**Comment by frol** [comment]
> I understand this part and the discussion about contract self upgrade. But I don't think we need resharing to implement the ticket's requirement.

I did not get it. Please, provide more context.

For announcements: We want admins to get top-level posts published as if the announcement account is posting them (admin will call some function on `devhub.contract` which will call `set("protocol.community.devhub.near/post/main", "blah")` on `social.near`). The comments are, indeed, don't need to have any special treatment.

For discussions: We want anyone (including admins) to just post on their behalf and then reshare ("repost") to `discussions.protocol.community.devhub.near` by calling some function on `devhub.near`.

**Comment by frol** [comment]
On a side note, I have just closed https://github.com/near/neardevhub-contract/issues/75 as I envision that community logo and other metadata can be set as `<community-handle>.communities.devhub.near` SocialDB profile metadata.

**Comment by ailisp** [comment]
> For announcements: We want admins to get top-level posts published as if the announcement account is posting them (admin will call some function on devhub.contract which will call set("protocol.community.devhub.near/post/main", "blah") on social.near). The comments are, indeed, don't need to have any special treatment.

We are on the same page here. And announcement doesn't need reshare.

> For discussions: We want anyone (including admins) to just post on their behalf and then reshare ("repost") to discussions.protocol.community.devhub.near by calling some function on devhub.near.

discussions needs reshare ("repost"), but given this ticket requirement all about announcement, let's address discussions on a separate ticket.

> On a side note, I have just closed https://github.com/near/neardevhub-contract/issues/75 as I envision that community logo and other metadata can be set as <community-handle>.communities.devhub.near SocialDB profile metadata.

Yes

**Comment by ailisp** [comment]
Opening a draft PR: https://github.com/near/neardevhub-contract/pull/79

**Comment by ori-near** [comment]
Is this ticket blocked?

**Comment by elliotBraem** [comment]
@ori-near #603 is ready for review

Notice that while commenting does not appear to be working in the preview, this is only because it is testnet and near.org's comment widget (which we are using) utilizes the near-query-api, and this indexer does not keep track of testnet comments. 
It will work in mainnet.
